### PR TITLE
Fix Confirmation Token to only work once - #2746

### LINKED ISSUE #2746: confirmation_token should work only once
### User Story

As a security person I want the confirmation_token to activate account to work only once.

### Why is this Needed?

For security reasons, email eaves-dropping, etc.

### Description

After registering and gettign activation link, clicking on activation link doesn't invalidate the token. This "bug" may shadowed by other, see #2745 

### Current Behavior

1. I've registered.
2. Got email with validation token.
3. Clicked the link and got response "Your account has been confirmed."
3. Clicked the again the link and got response "Your account has been confirmed."

### Expected Behavior

Second link should result in "Confirmation token has been already used."

### PR DESCRIPTION
**Pull request by iraline**
## References

* Closes #2746

## Objectives

Let the confirmation_token to activate account to work only once.

## Visual Changes
When trying to use some token already used, this is what happens

### CODE REVIEW

File: app/controllers/users/confirmations_controller.rb [hunk 1]
```diff
@@ -48,8 +48,15 @@ def show
       respond_with_navigational(resource) { render :show }
     elsif resource.errors.empty?
       set_official_position if resource.has_official_email?
-      resource.confirm # Last change: confirm happens here for people with passwords instead of af the top of the show action
-      set_flash_message(:notice, :confirmed) if is_flashing_format?
+
+      if resource.confirmed_at
```
Comment (javierm) [inline app/controllers/users/confirmations_controller.rb:52]: Comment automatically generated by Pronto.

Layout/TrailingWhitespace: Trailing whitespace detected. (https://rubystyle.guide#no-trailing-whitespace)

```suggestion
      if resource.confirmed_at
```

File: app/controllers/users/confirmations_controller.rb [hunk 2]
```diff
@@ -48,8 +48,15 @@ def show
...
-      set_flash_message(:notice, :confirmed) if is_flashing_format?
+
+      if resource.confirmed_at 
+        message = :already_confirmed
+      else
+        resource.confirm # Last change: confirm happens here for people with passwords instead of af the top of the show action
+        message = :confirmed
+      end
+     
+      set_flash_message(:notice, message) if is_flashing_format?
```
Comment (javierm) [inline app/controllers/users/confirmations_controller.rb:59]: Comment automatically generated by Pronto.

Layout/TrailingWhitespace: Trailing whitespace detected. (https://rubystyle.guide#no-trailing-whitespace)

```suggestion

      set_flash_message(:notice, message) if is_flashing_format?
```

**Comment by javierm** [comment]
@iraline Thanks for your contribution! :pray:

Before reviewing the pull request, I've just noticed the test "Account Edit email address" (line 53 in spec/system/account_spec.rb) is failing because with these changes it would be impossible to confirm an account after changing the email address. So that means this issue is trickier than we thought :thinking:.

Any ideas about how to handle this scenario?

**Comment by iraline** [comment]
@javierm I will analyze and let you know here. =)

**Review by javierm — COMMENTED** [review]
@iraline Thanks for the pull request!

I've made a comment about a possible way to do what Devise does by default so we don't have to change any internationalization keys. Let me know what you think!

File: app/controllers/users/confirmations_controller.rb [hunk 3]
```diff
@@ -48,8 +48,15 @@ def show
...
+
+      if !resource.confirmed_at || resource.unconfirmed_email
+        resource.confirm # Last change: confirm happens here for people with passwords instead of af the top of the show action
+        message = :confirmed
+      else
+        message = :already_confirmed
+      end
+
+      set_flash_message(:notice, message) if is_flashing_format?
       respond_with_navigational(resource) { redirect_to after_confirmation_path_for(resource_name, resource) }
```
[the remark this answers is not in the document]
Comment (iraline) [reply app/controllers/users/confirmations_controller.rb:60]: Yes, it makes sense. Thank you for your help. I will make these changes and update the tests.

**Review by javierm — APPROVED** [review]
@iraline Thanks a lot! :tada:

**Pull request merged**